mempool/src/core_mempool/mempool.rs

```diff
@@ -214,6 +216,21 @@ impl Mempool {
                     bucket,
                     insertion_to_block,
                 );
+
+                let use_case_label = match tracked_use_case {
+                    Some(UseCaseKey::Platform) => "platform".to_string(),
+                    Some(UseCaseKey::ContractAddress(addr)) => format!("contract_{:?}", addr),
+                    Some(UseCaseKey::Others) => "other".to_string(),
+                    None => "contract_other".to_string(),
+                };
+
+                counters::TXN_E2E_USE_CASE_COMMIT_LATENCY
+                    .with_label_values(&[
+                        &use_case_label,
```

Comment:
This is going to blow up the number of counters no?

Reply:
argh.

@sitalkedia - at any given block, we are exporting at most usecase_stats_num_top_to_track = 5 distinct values. (top is counted over 40 blocks, to be more stable)

so at any individual time it should be fine.

over time, these might change. from prometheus side, I don't think that is a problem - each forge test are new sets of counters.


But histogram will be exporting unchanged values too. (and using memory for them) Do you know of a way that will only keep track of new values, and export only those? Otherwise, I'll need to change this to contract_top1 / contract_top2 / ...

Reply:
Yeah, my concern is the number of dimensions being tracked over time. Not a prometheus export to say for sure if its a problem but need to understand the implication of this before landing this change.

Reply:
as is, it's also a problem that metrics that were created, but stopped being updated, continue being exported - as count is count from the start of the process.

I'll change to contract_top1 / ... , and log contract_top1 <-> address map, as I don't have a better suggestion

Reply:
I doubt our current prometheus db can support this churn. We are talking about 5*4blocks = 20 unique labels per second per peer. And, if we end up getting unique labels for minutes, that will definitely overwhelm the db. 

Forge is short-lived, when the node dies, it's labels go away with it. For mainnet, the validator will keep exporting the metric as long as it's alive, even if you record it once. 

cc: @geekflyer

Reply:
yeah agree with @ibalajiarun  . This is a problem, esp. the dynamic contract address label.
I think what we should here is this:

1. Put exporting this metric behind a feature flag for now. We can enable this feature flag for our internal nodes for a while to understand the implications better and before enabling it by default on external nodes.
2. We need to look into how to expire metrics from the process-internal metrics registry. I know that some libraries do support this, but it's not super common unfortunately. As Balaji rightly pointed out in its current state this would report lots of non-changing metrics that will ever increase the longer a node runs (leaky metrics registry lol), which is a problem for our TSDB but also creates extra network overhead etc.

Reply:
btw, have we thought about tracking this via logs or traces? If we for example have this behind a feature flag we could have one of the internal validators emit a log line for each observed TX. In terms of volume I think this should be fine as long we don't emit this for every single node.

Reply:
On 2. here's a related issue https://github.com/tikv/rust-prometheus/issues/336  (this is the lib we're using a the moment afaik).

Reply:
btw, just to add: I think our TSDB can possibly support this much cardinality/churn, but the issue is the unboundness of cardinality due to non-expiring metrics in the process registry, which amplifies the problem.

Reply:
I think we've been doing similar things for the module+function already?

Reply:
here we kinda have to have it for all - because only PFN that transaction is submitted to can compute latency correctly

But yeah, we have 2 fullnodes logging detailed countrs - but those are on module name, so republishing a contract under different address shows in the same counter. not sure if cardinalities there are causing any issues, or we need to revisit those counters - https://aptoslabs.grafana.net/goto/fxPdCAXSR?orgId=1

for here, I've basically changed cardinality to fixed set by using entry_user_top_1  ... entry_user_top_5, which should be enough